doc: ADR-0004 Exams IDA

Zacharis278 opened the pull request:
**Description:**

Intent to move features in this library to an IDA. I've intentionally left LTI details out of this for the most part. Since, from a code standpoint these two decisions are independent.

**JIRA:**


## [MST-1354](https://openedx.atlassian.net/browse/MST-1354)

MichaelRoytman commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+
+Decision
+--------
+A new repository and IDA will be created for the exam service. As service will handle all configuration and interfacing with third party proctoring tools. It will also be the source of truth for any state information about a learner's progress through exam setup, completion, and review flows.
```
`As service will handle all configuration and interfacing with third party proctoring tools.`

I'm not sure what this means. Could you clarify, please?

MichaelRoytman commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+Consequences
+------------
+
+* The display of learner attempt state in course outlines requires the LMS to directly depend on the proctoring service. This feature will be simplified to only contain information about the type of exam. Attempt state will be accessible within the section via the frontend-lib-special-exams library.
```
Can we tie this back to the decision `The LMS will not depend on this service.` so it's clear why this is a consequence of our decisions?

MichaelRoytman commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+
+* The display of learner attempt state in course outlines requires the LMS to directly depend on the proctoring service. This feature will be simplified to only contain information about the type of exam. Attempt state will be accessible within the section via the frontend-lib-special-exams library.
+
+* The LMS will need a new way to prevent rendering exam content while not activly in an exam that does not rely on the installation of this library.
```
```suggestion
* The LMS will need a new way to prevent rendering exam content while not actively in an exam that does not rely on the installation of this library.
```

MichaelRoytman replied:
This may be a nit, but you can also view exam content if you're not actively in the exam so long as the exam is past due and the exam content is available after the due date.

"while a learner is not eligible to view exam content" could be more generic and cover all the bases.

Zacharis278 commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+
+This led us to evaluate the need for these timed/proctored exam features and data models to be available alongside edx-plaform as a plugin. This library is currently coupled with the platform without a well-defined interface leading to many imports in both directions. We must use a runtime wrapper just to avoid the circular dependencies that exist in this approach today. However, the actual data  ...
+
+We believe we can create a well-defined interface that results in an exam solution that is best managed as its own service. The following proposed component diagram defines the relationship between platform components and the bounded context of timed exams.
```
good call, that would be more clear here

Zacharis278 commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+
+Decision
+--------
+A new repository and IDA will be created for the exam service. As service will handle all configuration and interfacing with third party proctoring tools. It will also be the source of truth for any state information about a learner's progress through exam completion and review flows.
```
I think I just had some words get mumbled together here lemme fix this sentence.

Zacharis278 commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+
+* We will need to maintain both proctoring implementations until we can migrate use away from this library and the vendors it specifically supports.
+
+
```
I don't think there ever was an alternative here. The #1 goal was to address the independent rate of change between this code and edx-platform, and ultimately not being tied to platform deployments. Another plugin doesn't really solve this. I probably need to underscore that better since a lot of this doc does just go off and talk about the IDA interface.

Zacharis278 commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+
+Context
+-------
+The process to release changes in this library to a production environment is both slow and involves multiple manual steps.
```
this is a good point, there's definitely something to be said about the complexity created by adding a new type of integration on top of the two we already have for proctortrack (somewhat generic) and PSI  (totally vendor specific). That definitely plays into the justification both time wise and in terms of code quality.

Zacharis278 commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,59 @@
+Waiting for the platform CI checks and deployment pipeline to run can take up to half a day. This adds significant turnaround
+to releasing even the smallest of changes.
+
+This led us to evaluate the need for these timed/proctored exam features and data models to be available alongside edx-plaform as a plugin. This library is currently coupled with the platform without a well-defined interface leading to many imports in both directions. We must use a runtime wrapper just to avoid the circular dependencies that exist in this approach today. However, the actual data  ...
```
I'm actually kind of wondering if referencing this is just unnecessary altogether.

Zacharis278 commented:
@MichaelRoytman I think I addressed most of your feedback. I tried to change some language around to focus more on why the desire to move this code to it's own service.

Also added some important context I totally missed. A major piece that enables this is the fact we can drop all the view logic that has moved to react.

ashultz0 commented on docs/decisions/0004-exam-ida.rst:
```diff
@@ -0,0 +1,64 @@
+
+* Existing proctoring vendors will remain functional using this library but will not by-default be implemented in the new IDA.
+
+* We will need to maintain both proctoring implementations until we can migrate use away from this library and the vendors it specifically supports.
```
I thought we were going to migrate everything to the IDA but let the IDA wrap the plugin so that we actually start using the IDA and its interface before we have an actual partner.

idegtiarov reviewed (commented):
I like this ADR since the legacy frontend is going to be removed from the platform, IDA for the special exam looks really great time to split the code into separate service.  

On the other hand, if the installation does not need special exam functionality it will be easily deployed without it that is great.

schenedx reviewed (approved):
👍
